crates/mun_hir/src/module_tree.rs

```diff
@@ -0,0 +1,227 @@
+            let mut module_id = root;
+            for path_segment in path_to_module_path(&relative_path)
+                .into_iter()
+                .map(Name::new)
+            {
+                module_id = match modules[module_id].children.get(&path_segment) {
+                    Some(id) => *id,
+                    None => {
+                        let child_module_id = modules.alloc(ModuleData {
+                            parent: Some(module_id),
+                            children: Default::default(),
+                            file: None,
+                        });
+                        modules[module_id]
+                            .children
+                            .insert(path_segment, child_module_id);
+                        child_module_id
+                    }
+                };
+            }
+
+            // Mark the found module with the current file id
+            let module = &mut modules[module_id];
+            if let Some(other_file) = module.file {
+                diagnostics.push(DuplicateModuleFile(module_id, vec![other_file, file_id]));
+            }
+
+            module.file = Some(file_id);
+        }
+
+        Arc::new(ModuleTree {
+            package,
+            diagnostics,
+            modules,
+            root,
+        })
+    }
+
+    /// Given a `RawVisibility` which describes the visibility of an item relative to a module into
+    /// a `Visibility` which describes the absolute visibility within the module tree.
+    pub(crate) fn resolve_visibility(
+        &self,
+        _db: &dyn DefDatabase,
+        original_module: LocalModuleId,
+        visibility: &RawVisibility,
+    ) -> Visibility {
+        match visibility {
+            RawVisibility::This => Visibility::Module(ModuleId {
+                package: self.package,
+                local_id: original_module,
+            }),
+            RawVisibility::Super => {
+                let parent_module_id = self[original_module].parent.unwrap_or(original_module);
+                Visibility::Module(ModuleId {
+                    package: self.package,
+                    local_id: parent_module_id,
+                })
+            }
+            RawVisibility::Package => Visibility::Module(ModuleId {
+                package: self.package,
+                local_id: self.root,
+            }),
+            RawVisibility::Public => Visibility::Public,
+        }
+    }
+
+    /// Returns the module that is defined by the specified `file`
+    pub fn module_for_file(&self, file: FileId) -> Option<LocalModuleId> {
+        self.modules.iter().find_map(|(idx, data)| {
+            if data.file == Some(file) {
+                Some(idx)
+            } else {
+                None
+            }
+        })
+    }
+}
+
+/// Given a path name, returns the module name
+fn normalize_module_name(name: Component) -> String {
+    name.as_str().replace("-", "_")
```

Comment:
Is `-` allowed in paths?

Reply:
It is on disk if thats whay you mean?

Reply:
My point is that if path names directly translate to module names, we should just mandate snake_case?

I think Rust originally allowed `-`, but later changed that. We can choose to just keep it consistent.

Reply:
I changed it so that it no longer normalized the path and just errors on paths with a `-`.